src/sync/once.rs

```diff
@@ -0,0 +1,147 @@
+    /// if this is the first time `call_once` has been called, and otherwise the routine will *not*
+    /// be invoked.
+    ///
+    /// This method will block the calling thread if another initialization routine is currently
+    /// running.
+    ///
+    /// When this function returns, it is guaranteed that some initialization has run and completed
+    /// (it may not be the closure specified).
+    pub fn call_once<F>(&self, f: F)
+    where
+        F: FnOnce(),
+    {
+        let lock = ExecutionState::with(|state| {
+            // Initialize the state of the `Once` cell if we're the first thread to try
+            if self.get_state(state).is_none() {
+                self.init_state(state, OnceInitState::Running(Rc::new(Mutex::new(false))));
+            }
+
+            let init = self.get_state(state).expect("must be initialized by this point");
+            let init_state = init.borrow();
+            trace!(state=?init_state, "call_once on cell {:p}", self);
+            match &*init_state {
+                OnceInitState::Complete(clock) => {
+                    // If already complete, just update the clock from the thread that inited
+                    let clock = clock.clone();
+                    drop(init_state);
+                    state.update_clock(&clock);
+                    None
+                }
+                OnceInitState::Running(lock) => Some(Rc::clone(lock)),
+            }
+        });
+
+        // If there's a lock, then we need to try racing on it to decide who gets to run their
+        // initialization closure.
+        if let Some(lock) = lock {
+            let mut flag = lock.lock().unwrap();
+            if *flag {
+                return;
+            }
+
+            trace!("won the call_once race for cell {:p}", self);
+            f();
+
+            *flag = true;
+            // We were the thread that won the race, so remember our current clock to establish
+            // causality with future threads that try (and fail) to run `call_once`. The threads
+            // that were racing with us will get causality through acquiring the `Mutex`.
+            ExecutionState::with(|state| {
+                let clock = state.increment_clock().clone();
+                *self
+                    .get_state(state)
+                    .expect("must be initialized by this point")
+                    .borrow_mut() = OnceInitState::Complete(clock);
+            });
+        }
+    }
+
+    /// Performs the same function as [`Once::call_once()`] except ignores poisoning.
+    // TODO we don't currently implement poisoning at all, so this just calls `call_once`
+    pub fn call_once_force<F>(&self, f: F)
+    where
+        F: FnOnce(&OnceState),
+    {
+        self.call_once(|| f(&OnceState));
+    }
+
+    /// Returns `true` if some [`Once::call_once()`] call has completed successfully.
+    pub fn is_completed(&self) -> bool {
```

Comment:
In the case where the state is `OnceInitState::Complete(C)`, you need to update the clock of the caller with `C`.